rust/src/flatfit/mod.rs

```diff
@@ -0,0 +1,126 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+
+const LOW_CAP: usize = 2;
+
+#[derive(Clone)]
+pub struct FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    front: usize,
+    back: usize,
+    size: usize,
+    buffer: RefCell<Vec<Item<Value>>>,
+    binop: PhantomData<BinOp>,
+}
+
+#[derive(Clone)]
+struct Item<Value> {
+    val: Value,
+    next: usize,
+}
+
+impl<Value> Item<Value> {
+    fn new(val: Value, next: usize) -> Self {
+        Self { val, next }
+    }
+}
+
+impl<Value, BinOp> FifoWindow<Value, BinOp> for FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new() -> Self {
+        Self {
+            front: 0,
+            back: 0,
+            size: 0,
+            buffer: RefCell::new(Vec::new()),
+            binop: PhantomData,
+        }
+    }
+    fn push(&mut self, val: Value) {
+        let capacity = self.buffer.borrow().capacity();
+        if self.size + 1 > capacity {
+            self.rescale(capacity * 2)
+        }
+        let mut buffer = self.buffer.borrow_mut();
+        self.size += 1;
+        let prev = self.back;
+        self.back = (self.back + 1) % self.size;
+        buffer[self.back].val = val;
+        buffer[prev].next = self.back;
+    }
+    fn pop(&mut self) {
+        let capacity = self.buffer.borrow().capacity();
+        if self.size > 0 {
+            self.front = (self.front + 1) % capacity;
+            self.size -= 1;
+            if self.size < capacity / 2 {
+                self.rescale(capacity / 2)
+            }
+        }
+    }
+    fn query(&self) -> Value {
+        let mut agg = Value::identity();
+        let mut buffer = self.buffer.borrow_mut();
+        if self.size > 0 {
+            let mut tracing_indices = Vec::new();
```

Comment:
We actually got a significant performance improvement by making the `tracing_indices` a member of the object itself. Its use didn't change, but it was able to avoid constantly constructing the temporary vector and allocating new memory for it. When it's a permanent member, it's also reusing the same memory on each query. The size of `tracing_indices` has a low bound, so it's not a large increase in memory use.

Reply:
Very good point. It is now fixed it in the new commit. I think I should try and do this for more of the data structures.

Reply:
In FlatFAT, the Rust implementation uses a `HashSet` to keep track of which unique parents have been updated. Time-complexity wise a set is probably as good as it gets but I'm wondering if there is a there is a data structure with less overhead for keys which are always indices. Do you have any advice? I will in any case change it to be a member.

Reply:
I think the `HashSet` is probably best. I believe you want a set to avoid putting in duplicate parents, which means the options are `HashSet` and `BTreeSet`. I believe the complexity of `BTreeSet` is worst and average-case O(log *n*) while `HashSet` is worst-case O(*n*), average-case O(1). But in this case, *n* is the number of parents of batch elements, which is guaranteed to be O(log *n*) of the window. So a worst-case O(*n*) with average-case O(1) is probably better than O(log *n*) worst and average-case.

What's important is that the size of the set is bounded by O(log *n*) of the window, which means Reactive's O(log *n*)  of the window guarantee is preserved.

As far as being specialized for `usize`, I'm not aware of anything.